File: bindings/php/tests/KdfTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * crypto_kdf - no official vector exists for this construction (D-45: no DSTU KDF standard or
 * reference implementation exists at all). Correctness here means determinism/distinctness,
 * matching the Rust crate's own property-test posture. Misuse: wrong-length master key/context
 * (infallible otherwise, D-66 - no rejection category, there is no tag to tamper with).
 */
final class KdfTest extends TestCase
{
    public function testDerivesADeterministicSubkey(): void
    {
        $masterKey = dstu_core_kdf_keygen();
        $a = dstu_core_kdf_derive_subkey($masterKey, 0, 'encrypt_');
        $b = dstu_core_kdf_derive_subkey($masterKey, 0, 'encrypt_');
        $this->assertSame($a, $b);
    }

    public function testDifferentSubkeyIdGivesADifferentSubkey(): void
    {
        $masterKey = dstu_core_kdf_keygen();
        $a = dstu_core_kdf_derive_subkey($masterKey, 0, 'context1');
        $b = dstu_core_kdf_derive_subkey($masterKey, 1, 'context1');
        $this->assertNotSame($a, $b);
    }

    public function testDifferentContextGivesADifferentSubkey(): void
    {
        $masterKey = dstu_core_kdf_keygen();
        $a = dstu_core_kdf_derive_subkey($masterKey, 0, 'context1');
        $b = dstu_core_kdf_derive_subkey($masterKey, 0, 'context2');
        $this->assertNotSame($a, $b);
    }

    public function testRejectsWrongLengthMasterKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_kdf_derive_subkey('too short', 0, 'context1');
    }

    public function testRejectsWrongLengthContext(): void
    {
        $masterKey = dstu_core_kdf_keygen();
        $this->expectException(\ValueError::class);
        dstu_core_kdf_derive_subkey($masterKey, 0, 'short');
    }

    public function testRejectsNegativeSubkeyId(): void
    {
        $masterKey = dstu_core_kdf_keygen();
        $this->expectException(\ValueError::class);
        dstu_core_kdf_derive_subkey($masterKey, -1, 'context1');
    }
}

File: bindings/java/src/test/java/ua/dstucrypto/dstucore/KdfTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * {@code crypto_kdf} - no official vector exists for this construction (D-45). Correctness here
 * means determinism/distinctness. Misuse: wrong-length master key/context.
 */
class KdfTest {
    @Test
    void deriveSubkeyIsDeterministic() {
        byte[] masterKey = Kdf.keygen();
        assertArrayEquals(
                Kdf.deriveSubkey(masterKey, 0, "encrypt_".getBytes()),
                Kdf.deriveSubkey(masterKey, 0, "encrypt_".getBytes()));
    }

    @Test
    void differentSubkeyIdGivesDifferentSubkey() {
        byte[] masterKey = Kdf.keygen();
        byte[] a = Kdf.deriveSubkey(masterKey, 0, "context1".getBytes());
        byte[] b = Kdf.deriveSubkey(masterKey, 1, "context1".getBytes());
        assertFalse(java.util.Arrays.equals(a, b));
    }

    @Test
    void differentContextGivesDifferentSubkey() {
        byte[] masterKey = Kdf.keygen();
        byte[] a = Kdf.deriveSubkey(masterKey, 0, "context1".getBytes());
        byte[] b = Kdf.deriveSubkey(masterKey, 0, "context2".getBytes());
        assertFalse(java.util.Arrays.equals(a, b));
    }

    @Test
    void wrongLengthMasterKeyIsRejected() {
        byte[] tooShort = "too short".getBytes();
        assertThrows(IllegalArgumentException.class, () -> Kdf.deriveSubkey(tooShort, 0, "context1".getBytes()));
    }

    @Test
    void wrongLengthContextIsRejected() {
        byte[] masterKey = Kdf.keygen();
        assertThrows(IllegalArgumentException.class, () -> Kdf.deriveSubkey(masterKey, 0, "short".getBytes()));
    }
}
